casper/src/main/resources/PoS.rhox

```diff
@@ -81,25 +89,56 @@ new  PoS,
-              new vaultCh, transferCh, userCh in {
+              new vaultCh, transferCh, userCh, processPayCh,
+                  depositCh,
+                  computeSum, totalStakeCh,
+                  computeDelta, rewardsDeltaCh,
+                  computeMergeMap, newRewardsCh in {
                 getUser!(*userCh) |
 
-                for(@userPk <- userCh) {
-                  deposit!(userPk, amount, *returnCh)
+                runMVar!(*stateCh, *processPayCh, *returnCh) |
+
+                for(@(rewards, bonds), payResultCh <- processPayCh;
+                    @userPk <- userCh) {
+                  deposit!(userPk, amount, *depositCh) |
+                  @ListOps!("fold", bonds.toList(), 0, *computeSum, *totalStakeCh) |
+                  contract computeSum(@(pk, (stake, _)), @acc, resultCh) = {
+                    resultCh!(acc + stake)
+                  } |
+
+                  for(@totalStake <- totalStakeCh) {
+                    @ListOps!("fold", bonds.toList(), {}, *computeDelta, *rewardsDeltaCh) |
+                    contract computeDelta(@(pk, (currentValidatorStake, _)), @acc, resultCh) = {
+                      resultCh!(acc.set(pk, (amount * currentValidatorStake) / totalStake))
+                    }
+                  } |
+
+                  for (@rewardsDelta <- rewardsDeltaCh) {
+                    @ListOps!("fold", rewardsDelta.toList(), rewards, *computeMergeMap, *newRewardsCh) |
+                    contract computeMergeMap(@(pk, rewardDelta), @newRewards, resultCh) = {
+                      resultCh!(newRewards.set(pk, newRewards.getOrElse(pk, 0) + rewardDelta))
+                    }
+                  } |
+
+                  for(@depositResult <- depositCh;
+                      @newRewards <- newRewardsCh) {
+                    payResultCh!((newRewards, bonds), depositResult)
+                  }
                 }
               }
             } |
 
             contract PoS(@"slash", @blockHash, return) = {
-              new userCh, invalidBlocksCh, getInvalidBlocks(`rho:casper:invalidBlocks`) in {
+              new userCh, invalidBlocksCh,
+                  getInvalidBlocks(`rho:casper:invalidBlocks`) in {
                 getInvalidBlocks!(*invalidBlocksCh) |
                 getUser!(*userCh) |
-                for (@invalidBlocks <- invalidBlocksCh; @bonds <- bondsCh; @userPk <- userCh) {
+                for (@invalidBlocks <- invalidBlocksCh; @(amounts, bonds) <- stateCh; @userPk <- userCh) {
                   new toBeSlashed in {
                     toBeSlashed!(invalidBlocks.getOrElse(blockHash, userPk)) |
                     for (@validator <- toBeSlashed) {
                       // TODO: Transfer to coop wallet instead of just simply setting bonds to 0
-                      bondsCh!(bonds.set(validator, (0, bonds.get(validator).nth(1)))) |
+                      stateCh!((amounts, bonds.set(validator, (0, bonds.get(validator).nth(1))))) |
```

Comment:
Does this say that when a validator gets slashed only its rewards get set to 0?

Reply:
I've set the rewards to zero. What else should be done at shashing?

Reply:
bonds?

Reply:
as in stake